Make `rustup show` output info in a more logical order

majaha opened the pull request:
I ran `rustup show` and noticed that its output is in the order:
installed toolchains -> installed targets for active toolchain -> active toolchain
I thought it would make sense to put "active toolchain" before " installed targets for active toolchain" so that they in order of decreasing scope:
installed toolchains -> active toolchain -> installed targets for active toolchain
 so that is what this patch does 🙂
Thoughts?

*Before:*
```
Matt@Matts-PC /c/d/p/r/rustup (show-order)> rustup show
Default host: x86_64-pc-windows-msvc
rustup home:  C:\Users\Matt\.rustup

installed toolchains
--------------------

stable-x86_64-pc-windows-msvc (default)
nightly-x86_64-pc-windows-msvc

installed targets for active toolchain
--------------------------------------

wasm32-unknown-unknown
x86_64-pc-windows-msvc

active toolchain
----------------

stable-x86_64-pc-windows-msvc (default)
rustc 1.67.1 (d5a82bbd2 2023-02-07)

Matt@Matts-PC /c/d/p/r/rustup (show-order)>
```
*After:*
```
Matt@Matts-PC /c/d/p/r/rustup (show-order)> RUSTUP_HOME=home CARGO_HOME=home home/bin/rustup show
Default host: x86_64-pc-windows-msvc
rustup home:  D:\programming\repos\rustup\home

installed toolchains
--------------------

stable-x86_64-pc-windows-msvc (default)
nightly-x86_64-pc-windows-msvc

active toolchain
----------------

stable-x86_64-pc-windows-msvc (default)
rustc 1.67.1 (d5a82bbd2 2023-02-07)

installed targets for active toolchain
--------------------------------------

wasm32-unknown-unknown
x86_64-pc-windows-msvc

Matt@Matts-PC /c/d/p/r/rustup (show-order)>
```

Fixes https://github.com/rust-lang/rustup/issues/1397

rbtcollins commented:
My gut reaction is that folk usually want the active toolchain information, so having it first or last is best; and last means they don't need to scan far for it or use head.

Neither first nor last will match your concept though.

Thanks for demonstrating the idea clearly though!

majaha commented:
Thanks for considering it!

I've fixed the tests I broke, amended the commit and commit message and force pushed. Apologies to anyone who pulled the original commit.

majaha commented:
There's always `rustup show active-toolchain [-v]`, although it is admittedly longer
```fish
Matt@Matts-PC /c/d/p/r/rustup (show-order)> rustup show active-toolchain
stable-x86_64-pc-windows-msvc (default)
Matt@Matts-PC /c/d/p/r/rustup (show-order)> rustup show active-toolchain -v
stable-x86_64-pc-windows-msvc (default)
rustc 1.67.1 (d5a82bbd2 2023-02-07)
```

majaha commented:
As an aside, would it be possible to reformat the reason the active toolchain was chosen? At the moment the floating `(default)` is a little misleading and mysterious if you don't already know what it means. And it visually matches how it looks in "installed toolchains" even though it means something different.  (My initial reaction to it was "Why is there a default active toolchain if there can only be one active toolchain? 🤔" but of course, that's not what it means 🙂).
```
active toolchain
----------------
stable-x86_64-pc-windows-msvc (default)
rustc 1.67.1 (d5a82bbd2 2023-02-07)
```
->
```
active toolchain
----------------
stable-x86_64-pc-windows-msvc
rustc 1.67.1 (d5a82bbd2 2023-02-07)
Reason: is the default toolchain
/or/
Reason: overridden by +toolchain on the command line
```

Also, there's some scope for code deduplication between `show_active_toolchain()` and `show()`

workingjubilee commented:
I feel like "\nReason:" is nondescript (reason for what? it's dissociated from the phrase "active toolchain" by multiple lines by that point) and the " (inline comment)" style is better, but I don't think " (default)" is a good explanation either.

majaha commented:
It's a little terse, I'll admit. I think because it's right next to the other parts of the active toolchain, you can work out what it's the reason for using context from the reason message. But I would be open to:
```
active toolchain
----------------
stable-x86_64-pc-windows-msvc
rustc 1.67.1 (d5a82bbd2 2023-02-07)
Reason it's active: it's the default toolchain
```

Either way, it's certainly more descriptive than the other lines! Perhaps it should really be:
```
active toolchain
----------------
Toolchain name: stable-x86_64-pc-windows-msvc
Compiler version: rustc 1.67.1 (d5a82bbd2 2023-02-07)
Reason it's active: it's the default toolchain
```

workingjubilee commented:
Hm. Why not reframe it and instead
```
active toolchain: default
----------------
stable-x86_64-pc-windows-msvc
rustc 1.67.1 (d5a82bbd2 2023-02-07)
```

rbtcollins commented:
What would happen if we put the reason on the same line as Active toolchain.

`Active toolchain (the default)`
`Active toolchain (selected by RUSTUP_TOOLCHAIN)`
...

(or something similar).

Having thought more about this I do like the idea of making targets subordinate to the toolchain. Perhaps like this:

```
RUSTUP_HOME=home CARGO_HOME=home home/bin/rustup show
Default host: x86_64-pc-windows-msvc
rustup home:  D:\programming\repos\rustup\home

installed toolchains
===================

stable-x86_64-pc-windows-msvc (default)
nightly-x86_64-pc-windows-msvc

active toolchain
================

channel: stable-x86_64-pc-windows-msvc
version: rustc 1.67.1 (d5a82bbd2 2023-02-07)
reason: default
targets: 
--------

wasm32-unknown-unknown
x86_64-pc-windows-msvc
```

majaha commented:
Yes, that's definitely an improvement!

Although I think we could do away with the blank lines after the `=====` and `-----` lines.

djc commented:
@majaha would you be able to rebase this and follow up on the discussion so far?

majaha commented:
Okay, I've rebased the old patch.

We're probably not ready to merge yet though, as we were still discussing and iterating on what we wanted to do. I'm going to have another think about what we want here.

majaha commented:
Okay, I've spent some more time thinking about the output format, incorporating some other's ideas, and this is what I've come up with:

```
$ rustup show
Default host: x86_64-pc-windows-msvc
rustup home:  C:\Users\Matt\.rustup

installed toolchains
====================
stable-x86_64-pc-windows-msvc (default)
nightly-x86_64-pc-windows-msvc

active toolchain
================
name: stable-x86_64-pc-windows-msvc
compiler: rustc 1.73.0 (cc66ad468 2023-10-03)
active because: overridden by +toolchain on the command line
targets:
--------
wasm32-unknown-unknown
x86_64-pc-windows-msvc
```

I thought about putting the active reason in the header like this:
```
active toolchain [by default]
================
```
but decided against it as the `rust-toolchain` file override reason can be arbitrarily long, and will probably wrap for some people, causing some headache with what to prevent the wrapped `====` line being ugly:
```
active toolchain [overridden by 'D:\foo\bar\baz\...\rust-toolchain']
================
```

We should also change the format of `rustup show active-toolchain` and split it into two lines:
```
$ RUSTUP_TOOLCHAIN=nightly rustup show active-toolchain
nightly-x86_64-pc-windows-msvc (environment override by RUSTUP_TOOLCHAIN)
```
⬇
```
$ RUSTUP_TOOLCHAIN=nightly rustup show active-toolchain
nightly-x86_64-pc-windows-msvc
active because: environment override by RUSTUP_TOOLCHAIN
```

rami3l commented:
> ```
> $ RUSTUP_TOOLCHAIN=nightly rustup show active-toolchain
> nightly-x86_64-pc-windows-msvc
> active because: environment override by RUSTUP_TOOLCHAIN
> ```

Indeed, breaking that into two lines might be a good idea. @majaha, it looks like we are moving in the right direction!

> ```
> active toolchain [overridden by 'D:\foo\bar\baz\...\rust-toolchain']
> ================
> ```

I still prefer using `active because:` here.

However, I'd like to add an `(active)` in the first section, which is a deliberate duplication but will hopefully make things clearer:

```
$ rustup show
Default host: x86_64-pc-windows-msvc
rustup home:  C:\Users\Matt\.rustup

installed toolchains
====================
stable-x86_64-pc-windows-msvc (active)
nightly-x86_64-pc-windows-msvc (default)

active toolchain
================
name: stable-x86_64-pc-windows-msvc
compiler: rustc 1.73.0 (cc66ad468 2023-10-03)
active because: overridden by +toolchain on the command line
targets:
--------
wasm32-unknown-unknown
x86_64-pc-windows-msvc
```

... and we might want to use `(active, default)` for the non-overridden state.

rami3l commented:
@rbtcollins @djc Instead of changing to `=` and `-` combined as proposed above, what if we keep the `-`, and add indentation to the mix (IIRC it should be 2 spaces to be coherent)?

```sh
<snip>

active toolchain
----------------
name: stable-x86_64-pc-windows-msvc
compiler: rustc 1.73.0 (cc66ad468 2023-10-03)
active because: overridden by +toolchain on the command line
targets:
  wasm32-unknown-unknown
  x86_64-pc-windows-msvc
```

Other than this I have no outstanding concerns at the moment.

majaha commented:
> I still prefer using `active because:` here.

Sorry, do you mean that you want it to be on the same line as the header next to `active toolchain`, or that you want it to be on it's own line underneath?

rami3l commented:
> > I still prefer using `active because:` here.
> 
> Sorry, do you mean that you want it to be on the same line as the header next to `active toolchain`, or that you want it to be on it's own line underneath?

A separate line underneath, just like in my snippets above.

majaha commented:
Ah, gotcha. Yeah, I think you're version looks quite clean!

djc commented:
This looks good to me. So can we get the code updated to match the proposed style?

majaha commented:
Okay, I'm working on this.

Here's a thought I had: Should we always have the `installed toolchains` and `active toolchain -> installed targets` lists, even if they have only a single entry? I think that would be clearer and more intuitive. If you haven't seen them before, it's not obvious that their non-presence is giving you information. It's too implicit, you have to go and carefully read the help page to pick up on it.

rami3l commented:
@majaha As a quick reminder, with the currently proposed style, the output should be like:

```console
$ RUSTUP_TOOLCHAIN=nightly rustup show active-toolchain
nightly-x86_64-pc-windows-msvc
active because: environment override by RUSTUP_TOOLCHAIN
```

_(From https://github.com/rust-lang/rustup/pull/3225#issuecomment-1782058807)_

---

... with `(active)` and `(default)` in `installed toolchains`:

```console
$ rustup show
Default host: x86_64-pc-windows-msvc
rustup home:  C:\Users\Matt\.rustup

installed toolchains
--------------------
stable-x86_64-pc-windows-msvc (active)
nightly-x86_64-pc-windows-msvc (default)

active toolchain
----------------
name: stable-x86_64-pc-windows-msvc
compiler: rustc 1.73.0 (cc66ad468 2023-10-03)
active because: overridden by +toolchain on the command line
targets:
  wasm32-unknown-unknown
  x86_64-pc-windows-msvc
```

_(From https://github.com/rust-lang/rustup/pull/3225#issuecomment-1793316750)_

---

Also, `(active, default)` should be used if the default toolchain is active.

_(From https://github.com/rust-lang/rustup/pull/3225#issuecomment-1793314607)_

majaha commented:
I've uploaded the new patch, which includes the changes and a bit of a refactoring:
1. Change the output format of `rustup show`, to be in a more logical
   order and have more pleasing formatting
2. Update the formatting of `rustup show active-toolchain` and
   `rustup toolchain list` to match the new `rustup show` format.
3. `rustup +nightly show` will no longer install the nightly toolchain
   if it isn't already installed. Ditto for
   `rustup show active-toolchain`. Fixes https://github.com/rust-lang/rustup/issues/1397
4. Fix a bug where the RUSTUP_TOOLCHAIN environment variable took
   priority over the +toolchain command line option, and add a test for
   override priority.
5. `rustup default` no longer errors when there is no default toolchain
   configured, it prints a message instead.
6. Update the docs, and fix an issue where they incorrectly stated
   that `rust-toolchain.toml` config files couldn't specify custom
   names. (Text last touched here, see commit message: https://github.com/rust-lang/rustup/commit/7bfbd3b96c29898de001cbda7730fa440d35bd3b)

djc commented:
@majaha this sounds great, except I'd like to request basically one commit per bullet point, which should make it a lot easier to review this. Do you think that would be feasible?

majaha committed:
Update format of `show` and `show active-toolchain`

This changes the format of `rustup show` to be in a more logical order,
and changes the format of `rustup show active-toolchain` to match.
Also, as suggested in a comment, these commands will no longer install
the active toolchain if it is not already installed, as they now call
`cfg.find_active_toolchain()` instead of
`cfg.find_or_install_active_toolchain()`. This fixes
https://github.com/rust-lang/rustup/issues/1397

majaha committed:
Update format of `toolchain list`

Update the format of `rustup toolchain list` to have a similar flavour
to the new `rustup show` format.

majaha committed:
Make `rustup default` not error if no default

`rustup default` will no longer error and print to stderr if there is no
default toolchain configured. It will print to stdout and exit
successfully instead.

majaha committed:
Fix doc error with `rust-toolchain.toml` custom TC

Fix an issue where the documentation erroneously states that
`rust-toolchain.toml` config files can't specify custom toolchain names.

This language seems to be very old, ultimately inherited from the
original commit that added this feature, and is no longer true:
https://github.com/rust-lang/rustup/commit/107d8e5f1ab83ce13cb33a7b4ca0f58198285ee8#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R330

majaha commented:
Ok, I've split out as best as I can. The first one is still quite large, but I can't really split it much further without meticulously reimplementing the bugs that were fixed. The commit message should help though; start by reading through the changes to `OverrideCfg`, and noting that `OverrideReason` has turned into `ActiveReason`, and the rest mostly follows from there.

djc commented on src/config.rs:
Here, I'd like a separate commit for the refactoring that splits the `match` for `toolchain_name` out of the `Self` initializer.

rami3l replied:
I second this.

djc commented on src/config.rs:
```diff
@@ -96,17 +96,20 @@ impl<T: Into<String>> From<T> for OverrideFile {
+// Represents the reason why the active toolchain is active.
 #[derive(Debug)]
-pub(crate) enum OverrideReason {
+pub(crate) enum ActiveReason {
```
This rename should be in a separate commit.

rami3l replied:
@djc I'm not so sure about that... Although it could possibly make the PR easier to review, the intermediate state that you're looking for might have inconsistent semantics (that is, a "new" type/fn name with an "old" implementation which cannot match that name). IMHO it's like intentionally introducing bugs into the codebase 🤔

djc replied:
I generally don't think this is a big issue? Like, it's only within the context of a few commits (that are part of the same PR), after which the branch is in a consistent state again.

majaha replied:
My instinct is that one semantic change should be in one commit, the code should compile and pass tests at each commit, and that artificially small commits aren't helpful. That's what I'm used to from my experience working on different projects. I don't really see any value in having the branch in an inconsistent state to begin with.

djc commented on src/cli/common.rs:
```diff
@@ -491,22 +492,24 @@ fn print_toolchain_path(
+            if let Ok(Some((LocalToolchainName::Named(toolchain), reason))) =
+                cfg.find_active_toolchain(&cwd)
+            {
+                if !matches!(reason, crate::config::ActiveReason::Default) {
```
Style nit: let's import `ActiveReason` here, and use a `match` expression` instead of `matches!()`. Also please avoid type annotations (here and elsewhere) -- I think this one is probably unnecessary.

djc commented on src/cli/common.rs:
```diff
@@ -491,22 +492,24 @@ fn print_toolchain_path(
-        .iter()
-        .map(Into::into)
-        .collect::<Vec<_>>();
+    let toolchains: Vec<ToolchainName> = cfg.list_toolchains()?;
```
This doesn't need a type annotation, please remove it.

djc commented on src/config.rs:
```diff
@@ -115,59 +118,157 @@ impl Display for OverrideReason {
+    name: LocalToolchainName,
+    reason: &ActiveReason,
+) -> Result<Toolchain<'a>> {
+    match Toolchain::new(cfg, name.clone()) {
```
Please rewrite this to reduce rightward drift:

```rust
match Toolchain::new(cfg, name.clone()) {
     Err(RustupError::ToolchainNotInstalled(_)) => {},
     result => return Ok(result?),
}

..
```

djc commented on src/config.rs:
```diff
@@ -443,105 +544,62 @@ impl Cfg {
     }
 
-    pub(crate) fn find_override(
+    pub(crate) fn find_active_toolchain(
```
I think this rename should probably in a commit by itself?

rami3l replied:
@djc I'm not so sure about that... Although it could possibly make the PR easier to review, the intermediate state that you're looking for might have inconsistent semantics (that is, a "new" type/fn name with an "old" implementation which cannot match that name). IMHO it's like intentionally introducing bugs into the codebase 🤔

djc reviewed (commented):
I'm sorry, but I do think the first commit should be split up -- as is, this is very hard to review. I reviewed that commit only so far, suggesting some ways to make it smaller. Overall the direction looks great!

majaha committed:
Pull out `ensure_installed()`

majaha committed:
Pull out `new_toolchain_with_reason()`

majaha committed:
Change find_override to find_active_toolchain

Changes the model from "find an override to the default toolchain" to
"find the active toolchain, which may be the default toolchain".

majaha committed:
Make settings file allow multiple borrows

majaha committed:
Make find_override_from_dir_walk return OverrideCfg

majaha committed:
Pull match statement out in OverrideCfg::from_file()

majaha committed:
Redesign OverrideCfg to be more type-driven

Redesigns the `OverrideCfg` type so that validity is represented in
the type system, i.e. "Parse, don't validate". Probably fixes some
subtle bugs when no toolchain is named in a rust-toolchain.toml,
implicitly selecting the default toolchain, but the default toolchain
isn't displayed as active by `rustup list toolchain` and the like.

Also fixes a bug where the `RUSTUP_TOOLCHAIN` erroneously had priority
over a `+toolchain` command line override.

majaha commented:
Looks like a random test failure. Can we run it again?

rami3l commented on src/config.rs:
According to our [API guidelines](https://rust-lang.github.io/api-guidelines/naming.html), `get_*` should probably be replaced with `into_`.

rami3l commented on src/config.rs:
```diff
@@ -443,106 +544,72 @@ impl Cfg {
-                Ok(())
-            })?;
+    ) -> Result<Option<(LocalToolchainName, ActiveReason)>> {
+        if let Some((override_config, reason)) = self.find_override_config(path)? {
```
How about using `Ok(if let ...)` here instead of `if let ... { Ok(...) } else { Ok(...) }`?

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -11,6 +11,7 @@ use clap::{
 };
 use clap_complete::Shell;
 
+use crate::config::new_toolchain_with_reason;
```
This import line should be merged to the group right below.

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -1078,133 +1088,112 @@ fn show(cfg: &Cfg, m: &ArgMatches) -> Result<utils::ExitCode> {
+
+    let active_targets: Vec<ComponentStatus> = active_toolchain_name
+        .and_then(|atn| match atn {
+            ToolchainName::Official(desc) => DistributableToolchain::new(cfg, desc.clone()).ok(),
```
Is there a particular reason for deviating from the original `DistributableToolchain::try_from(&at.0)` here?

majaha replied:
Yes: It's because `DistributableToolchain::try_from()` takes a `Toolchain`, which represents an installed toolchain on disk. This is what is returned by `find_or_install_active_toolchain()`.
The new code works with `ToolchainName`, which is what's returned by `find_active_toolchain()` above. This represents the name of a toolchain, which doesn't necessarily have to be installed.
Using `find_active_toolchain()` rather than `find_or_install_active_toolchain()` is what makes `rustup show` no longer install the active toolchain, if it isn't already installed.

rami3l replied:
@majaha I see. Thanks for your detailed explanation!

Should `CustomToolchainName` be considered as well though? In this case I think it's okay to assume the toolchain is installed.

majaha replied:
This code is for determining the list of targets that are installed for the active toolchain. Rustup only has the ability to query this for Official toolchains, not custom ones, as far as I know. That's what the previous code does, and that's what the `DistributableToolchain` is used for.

To be honest, the variable names could be a bit clearer. And I might put `CustomToolchainName` in the match statement directly. Explicit is better than implicit and all that. I'll change this quickly.

rami3l commented on src/cli/common.rs:
```diff
@@ -461,74 +462,72 @@ pub(crate) fn list_installed_components(distributable: DistributableToolchain<'_
+            String::new()
+        };
+        let status_str = match (is_default, is_active) {
+            (true, true) => " (default, active)",
```
As stated in https://github.com/rust-lang/rustup/pull/3225#issuecomment-1793314607 I personally think it's better to use `(active, default)` since `active` will be more important info that I'll look for in this output.

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -1227,31 +1218,27 @@ fn show(cfg: &Cfg, m: &ArgMatches) -> Result<utils::ExitCode> {
         }
+        None => writeln!(
+            process().stdout().lock(),
+            "no default toolchain is configured"
```
You might want to say `active` instead of `default` here?

rami3l commented:
@majaha Sorry for the late review! I'm finally gaining bandwidth since a few days ago 💦

majaha commented:
No worries, hope you enjoyed your break! I've addressed your comments and uploaded the new version.

rami3l commented:
@majaha Looks like the `cargo fmt` check has failed, would you mind updating your code again?

majaha commented:
Ah sorry, there's always something. Should be fixed now.

rami3l commented:
~~@djc I've reviewed this PR and it looks ready in its current state.~~
~~I'd like to merge this PR soon and return to #3501, if that sounds okay to you.~~

Update: After my discussion with @rbtcollins last night we decided to postpone this change to the next milestone to better integrate this PR with the massive changes to be made in #3367 and #3596.

rbtcollins commented:
Hi, I know I'm coming to the code aspects of this late.

This PR is not focused on changing the output; it is instead:
- changing how we model toolchain selection internally
- changing the semantics for config setting file operations
- refactoring the show function to decompose it somewhat
- and changing the ordering

And perhaps more I've missed.

I'd like to take each of these separately to allow decent discussion about the merits of each, please.

djc commented:
> Update: After my discussion with @rbtcollins last night we decided to postpone this change to the next milestone to better integrate this PR with the massive changes to be made in #3367 and #3596.

I usually prefer not to consider Discord load-bearing in terms of project decisions. What is the rationale for postponing this change? #3367 and #3596 are large changes but are also mostly mechanical in nature.

rbtcollins commented:
> What is the rationale for postponing this change? #3367 and #3596 are large changes but are also mostly mechanical in nature.

https://github.com/rust-lang/rustup/pull/3225#issuecomment-1913593183 highlights my concerns.

majaha commented:
I'm open to discussing the changes. Do you have any specific issues with the code? I also agree that there's no need to postpone this beyond https://github.com/rust-lang/rustup/pull/3367 and https://github.com/rust-lang/rustup/pull/3596, they seem to be a largely separate issue - not to mention that this PR has been finished for some time now, it should really go in first.

djc commented:
> I'd like to take each of these separately to allow decent discussion about the merits of each, please.

Aren't these already factored into separate commits within this PR?

rami3l commented:
I still want to merge this, probably before v1.28.0 now that we definitely have a v1.27.1 to launch. I'll try to find some time to give it look once again. My apologies to @majaha for your long wait!

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -1078,133 +1087,106 @@ fn show(cfg: &Cfg, m: &ArgMatches) -> Result<utils::ExitCode> {
+    let active_toolchain_targets: Vec<TargetTriple> = active_toolchain_name
+        .and_then(|atn| match atn {
+            ToolchainName::Official(desc) => DistributableToolchain::new(cfg, desc.clone()).ok(),
+            ToolchainName::Custom(_) => None,
```
```suggestion
            // So far, it is not possible to list targets for a custom toolchain.
            ToolchainName::Custom(_) => None,
```

rami3l reviewed (approved):
v1.27.1 took us longer than expected, but I didn't forget about this PR. I just had another look, and I'm again confirmed that we can include this into v1.28.0.

majaha commented:
Sounds good. Let me know if I can help with anything.

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -1078,133 +1087,107 @@ fn show(cfg: &Cfg, m: &ArgMatches) -> Result<utils::ExitCode> {
+                )?;
+                writeln!(t.lock(), "name: {}", active_toolchain.name())?;
+                writeln!(t.lock(), "compiler: {}", active_toolchain.rustc_version())?;
+                writeln!(t.lock(), "active because: {}", active_reason.to_string())?;
```
```suggestion
                writeln!(t.lock(), "active because: {}", active_reason)?;
```

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -1227,31 +1212,24 @@ fn show(cfg: &Cfg, m: &ArgMatches) -> Result<utils::ExitCode> {
-            if let Some(reason) = reason {
+    match cfg.find_active_toolchain(&cwd)? {
+        Some((toolchain_name, reason)) => {
+            let toolchain = new_toolchain_with_reason(cfg, toolchain_name.clone().into(), &reason)?;
```
```suggestion
            let toolchain = new_toolchain_with_reason(cfg, toolchain_name.clone(), &reason)?;
```

djc commented on src/cli/download_tracker.rs:
```diff
@@ -181,7 +181,7 @@ impl DownloadTracker {
                         let remaining = content_len - self.total_downloaded;
                         let eta_h = Duration::from_secs(if speed == 0 {
-                            std::u64::MAX
+                            u64::MAX
```
Should really inline the clippy fixes into the commit that originated them (maybe try https://github.com/tummychow/git-absorb).

rami3l replied:
@djc TIL about git-absorb! I will give it a try on this one. However the std::u64::MAX wasn't introduced in this patch at all, so it cannot be squashed.

rami3l replied:
Never mind, looks like it has been done elsewhere, so I can safely drop it.

rami3l replied:
@djc Thanks a lot for the recommendation! `git-absorb` literally saved my day :)

majaha commented:
Nice, well done on merging this guys!
